### ISSUE TITLE
a change due to `GAP.prompt()`

### ISSUE THREAD

**Issue by ThomasBreuer** [oscar-system/GAP.jl#461]
In a GAP session started from Julia with `GAP.prompt()`, one gets the following.
```
gap> JuliaModule( "GAP" );
<Julia: GAP>
```
In a GAP session started in the traditional way, one gets the following.
```
gap> JuliaModule( "GAP" );
<Julia: Main.GAP>
```
(I just stumbled over this because it occurs in one of the test files for JuliaInterface.)

**Comment by fingolfin** [comment]
What exactly do you mean with "startedin the traditional way"? I.e. what are you doing?

If I do `$(julia -e 'import GAP; print(GAP.gap_exe())')` I get the first behaviour you list!

And indeed, this looks like an improvement to me; the "old" way of loading julia-in-gap necessarily was not following exactly the same steps as gap-in-julia did; one of the things we had to deal with is that we might not be able to do `using GAP` inside Julia due to the usual issues with that... so we have this hack in JuliaInterface:
```
if not IsBound( __JULIAINTERNAL_LOADED_FROM_JULIA ) then
    JuliaEvalString( "Base.MainInclude.include(Base.find_package(\"GAP\"))" );
fi;
```
That "emulates" the behaviour of `using GAP` but without requiring GAP.jl to be in the user's active project / environment. But this emulation is not perfect, and so the name of the resulting module ends up being `Main.GAP` instead of `GAP`.

**Comment by ThomasBreuer** [comment]
@fingolfin 

1. The traditional way to start GAP is to call the its executable, where the root directories are specified via the `-l` option. If this is no longer allowed in the presence of GAP-Julia integration then we have to state this in the documentation (of JuliaInterface, say) and to tell the reasons (like what you have sketched); is this the case?

2. As I wrote, the output `<Julia: Main.GAP>` occurs in one of the test files for JuliaInterface. The fact that the CI tests do not complain about it means that either the test in question does not run or the test setup is not up to date. If the answer to 1. is yes then I can try to find out what is wrong here.

**Comment by fingolfin** [comment]
Actually, I'd say the traditional way for 99% of all users is to run `bin/gap.sh` provided with GAP, not to invoke the gap executable directly with `-l` options...

That said, it is still "allowed" to do so, but fact is that there will be subtle difference when doing this, difference that may be difficult to remove.

The reason the CI tests complain is because the CI tests use bin/gap.sh. and so they happily execute the test case correctly

**Comment by ThomasBreuer** [comment]
@fingolfin I do not understand your last sentence:
In pull request #468, I have changed the test output to `<Julia: GAP>`, which is the correct result when GAP gets started from Julia. Travis complains (see https://travis-ci.com/github/oscar-system/GAP.jl/jobs/337445417) that the output should be `<Julia: Main.GAP>`.
My interpretation is that the test runs (which is good) but that it does not run in the expected way (which is bad).

**Comment by fingolfin** [comment]
Sorry, my bad, I totally misunderstood what you wrote. So, yeah, you are right: this is really weird.

When I run the test suite locally (after starting GAP via `./deps/build/gap-4.11.0-julia-1.4/bin/gap.sh`), I see this:
```
gap> TestPackage("juliainterface");
Architecture: x86_64-apple-darwin18.7.0-julia64-kv7-v1.4

testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/JuliaToGAP.tst
     200 ms (199 ms GC) and 48.8KB allocated for JuliaToGAP.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/adapter.tst
     201 ms (194 ms GC) and 249KB allocated for adapter.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/calls.tst
     358 ms (195 ms GC) and 438KB allocated for calls.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/convert.tst
     201 ms (191 ms GC) and 449KB allocated for convert.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/help.tst
    2277 ms (252 ms GC) and 335MB allocated for help.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/import.tst
     202 ms (197 ms GC) and 121KB allocated for import.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/juliainterface01.tst
########> Diff in /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/juliainterface01.tst:92
# Input is:
gapmodule:= JuliaModule( "GAP" );
# Expected output:
<Julia: Main.GAP>
# But found:
<Julia: GAP>
########
    1432 ms (211 ms GC) and 265KB allocated for juliainterface01.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/juliainterface02.tst
     409 ms (203 ms GC) and 99.8KB allocated for juliainterface02.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/utils.tst
     203 ms (200 ms GC) and 230KB allocated for utils.tst
testing: /Users/mhorn/Projekte/OSCAR/GAP.jl/pkg/JuliaInterface/tst/wrapper.tst
     198 ms (198 ms GC) and 64.2KB allocated for wrapper.tst
-----------------------------------
total      5681 ms (2040 ms GC) and 337MB allocated
              1 failures in 1 of 10 files

#I  Errors detected while testing

```

Weird. I would now starting wondering if this could be an issue on Travis, but my experimental CI test runner for GitHub Actions behave the same as Travis (i.e., they pass, even though I am using `gap.sh` here, too... Odd

**Comment by ThomasBreuer** [comment]
@fingolfin Thanks for the confirmation.
The outcome could have been worse,
cf. http://www.math.rwth-aachen.de/~Thomas.Breuer/eyetest.html

**Comment by ThomasBreuer** [comment]
The reason for the problem is that it makes a difference whether one first starts GAP and then runs the tests from the GAP prompt or whether one runs the tests via a file mentioned on the command line.
In the former case, the GAP input `JuliaModule("GAP")` yields the output `<Julia: GAP>`,
whereas one gets `<Julia: Main.GAP>` in the latter case.
(Note that the Travis tests are run using the GAP command line.
Also note that currently one gets a Julia warning
```
WARNING: using GAP.GAP in module Main conflicts with an existing identifier.
ERROR: LoadError: importing GAP into Main conflicts with an existing identifier
```
when one reads a file via the GAP command line, calls `LoadPackage("JuliaInterface");` inside this file, and then does *not* force GAP to terminate at the end of this file.)

The point is that currently the initializations from GAP.jl are not yet completed when the files from the command line are processed. Thus this problem should be solved in the context of issue #518.
